doc: cmake overhaul

gmarull opened the pull request:
Summary:

- Prefix with `_` all Sphinx-related directories, this makes clear what is a docset or not
- Move Sphinx extensions down to `doc` (they are not scripts and are only used by docs)
- Created `external_content` extension. The idea is that Sphinx handles external content itself (no CMake copy anymore).
- Moved to `doxyrunner`
- CMake cleanup (first round, more will come)

ru-fu commented on doc/nrf/conf.py:
```diff
@@ -103,18 +105,29 @@
+doxyrunner_outdir = NRF_BUILD / "doxygen"
+doxyrunner_fmt = True
+doxyrunner_fmt_vars = {
+    "NRF_BASE": str(NRF_BASE),
```
Wouldn't it make more sense to keep doing that replacement with cmake, instead of repeating all required variables here?
Also, if we do it like this here, we should probably use a different file name than nrf.doxyfile.in, since that implies that you do the "normal" cmake replacement.

gmarull replied:
This is how the plugin works, CMake doesn't take care of Doxygen anymore (unless we want to provide a standalone target, done e.g. in Zephyr). The technical discussions can be found here: https://github.com/zephyrproject-rtos/zephyr/pull/33934

ru-fu replied:
From what I understand, the plugin takes a doxyfile as input, and optionally does some replacements in the file before using it.
So why would we list the replacements manually if cmake can do them for free, and then we just provide the output doxyfile to the plugin?

gmarull replied:
As mentioned we already discussed that. The concept behind is that Sphinx does not depend on CMake for operating, so one could potentially just invoke `sphinx-build` in the future.

carlescufi replied:
This is temporary until we add a `wet docbuild` command that will encapsulate this. The usage of cmake for building the doc will slowly fade away until it's replaced by west and sphinx itself.

mbolivar-nordic commented on doc/nrf/known-warnings.txt:
```diff
@@ -1,5 +1,5 @@
-.*Duplicate C declaration.*\n.*'\.\. c:function:: .*bt_mesh_light_ctl_set.*'.*
-.*Duplicate C declaration.*\n.*'\.\. c:function:: .*bt_mesh_light_temp_set.*'.*
-.*Duplicate C declaration.*\n.*'\.\. c:struct:: bt_dfu_smp_rsp_state'.*
+.*Duplicate C declaration.*\n.*'\.\. c:.*:: .*bt_mesh_light_ctl_set.*'.*
```
Any plans to root-cause these and see if we're masking real problems at some point?

mbolivar-nordic reviewed (approved):
I just gave these a quick skim, but they seem OK to me

tejlmand commented on doc/CMakeLists.txt:
```diff
@@ -330,23 +363,39 @@ add_custom_target(
+add_dependencies(nrfxlib-html-all nrf-html-all)
+
 add_custom_target(
   build-all
```
Should we add `ALL` here, so that users may simply run `ninja` / `make` and get the complete doc set built.

Currently, users have to run `ninja build-all`
```suggestion
  build-all ALL
```

gmarull replied:
changed, thanks!

tejlmand commented on doc/nrf/doc_build.rst:
```diff
@@ -148,25 +148,25 @@ To clean the build folders for the Zephyr documentation:
 .. code-block:: console
 
-   ninja clean-zephyr
+   ninja zephyr-clean
```
How important is it to be able to clean each individual step ?
Wouldn't it be safer to use the already existing target `ninja clean` and then do a complete clean instead.

ru-fu replied:
It's required - we don't want to build for 20 minutes every time we need a clean nrf or nrfxlib doc build.

mbolivar-nordic replied:
>  Wouldn't it be safer to use the already existing target `ninja clean` and then do a complete clean instead.

Safer, but it is so very, very slow. I think it's nice to be able to clean individual pieces and not lose the other docsets.

gmarull replied:
I have setup `ninja clean` so that it cleans all artifacts, but individual cleans are still available.

tejlmand reviewed (commented):
Some observations, nothing blocking.

gmarull commented on doc/CMakeLists.txt:
```diff
@@ -330,23 +363,39 @@ add_custom_target(
-  COMMAND ${CMAKE_MAKE_PROGRAM} nrfxlib-html
-  USES_TERMINAL
+  DEPENDS
+    nrf-html-all
```
While it's redundant, I think it's more clear for people adding new docsets.

tejlmand replied:
accepted.

gmarull commented on .github/workflows/docbuild.yml:
```diff
@@ -58,7 +58,7 @@ jobs:
           mkdir -p _build && cd _build
-          cmake -GNinja ../nrf/doc -DEXTRA_SPHINXOPTS="-W"
+          cmake -GNinja ../nrf/doc -DSPHINXOPTS_EXTRA="-W"
           ninja build-all
```
done, thanks

gmarull commented on doc/CMakeLists.txt:
```diff
@@ -106,17 +104,9 @@ add_dependencies(zephyr html)
-    COMMAND ${CMAKE_COMMAND} -E remove ${TARGET_BINARY_DIR}/*.warnings
-    COMMAND ${CMAKE_COMMAND} -E remove ${TARGET_BINARY_DIR}/last_doxy_run_tstamp
-    # Remove generated HTML as well
+    COMMAND ${CMAKE_COMMAND} -E remove_directory ${CMAKE_BINARY_DIR}/${target}
```
Thanks for this suggestion. I've added support for `ADDITIONAL_CLEAN_FILES` in the last commit, can you review?

ru-fu reviewed (approved):
No complete check from my side, but I verified that the build works locally.

gmarull commented:
@tejlmand As I'll do some more CMake cleanups soon, let's do another review then. I have addressed the issues/suggestion you reported, can be tweaked later if they can be improved.
